node/node-sdk.md

```diff
@@ -0,0 +1,401 @@
+
+The SDK steers developers to write their workflows and activities in TypeScript but vanilla JS is also supported.
+All examples in this proposal are written in TypeScript.
+
+## Table Of Contents
+<!-- vim-markdown-toc GFM -->
+
+* [Activities](#activities)
+  * [Example](#example)
+  * [`ActivityContext`](#activitycontext)
+  * [Registration](#registration)
+  * [Payloads](#payloads)
+* [Workflows](#workflows)
+  * [Interface](#interface)
+  * [Implementation](#implementation)
+    * [Calling activities](#calling-activities)
+    * [Activity Options](#activity-options)
+      * [Global Activity Configuration](#global-activity-configuration)
+      * [Per activity override](#per-activity-override)
+      * [Activites with no type definitions](#activites-with-no-type-definitions)
+  * [Signals](#signals)
+  * [Queries](#queries)
+  * [Limitations](#limitations)
+  * [Logging](#logging)
+  * [Sessions](#sessions)
+  * [Deterministic Time and Random](#deterministic-time-and-random)
+  * [Error Handling](#error-handling)
+  * [Cancellation](#cancellation)
+    * [Example](#example-1)
+    * [`CancellationScope.run()`](#cancellationscoperun)
+  * [Long Histories](#long-histories)
+* [Project Structure](#project-structure)
+  * [References](#references)
+
+<!-- vim-markdown-toc -->
+
+## Activities
+
+Activity implementations are just JS functions, they run in the default NodeJS runtime, along with the Temporal SDK's worker, and can import arbitrary npm packages.
+
+### Example
+```ts
+// activities/greeter.ts
+
+export async function greet(name: string) {
+  return `Hello, ${name}!`;
+}
+```
+
+### `ActivityContext`
+
+Activities can get their `ActivityContext` by using `ActivityContext.current()`.
+
+This is implemented using [`AsyncLocalStorage`](https://nodejs.org/api/async_hooks.html#async_hooks_class_asynclocalstorage) which was introduced in `nodejs v13.10.0, v12.17.0`.
+(We can support earlier node versions by using the building blocks in [`async_hooks`](https://nodejs.org/api/async_hooks.html)).
+
+```ts
+// activities/http.ts
+
+import { ActivityContext } from 'temporal-sdk';
+import axios from 'axios';
+
+export async httpGet(url: string): Promise<string> {
+  const response = await axios.get(url, {
+    onDownloadProgress: (progressEvent) => {
+      const total = parseFloat(progressEvent.currentTarget.responseHeaders['Content-Length']);
+      const current = progressEvent.currentTarget.response.length;
+      const progress = Math.floor(current / total * 100);
+      ActivityContext.current().heartbeat(progress);
+    },
+  });
+  return response.data;
+}
+```
+
+### Registration
+Activities are automatically registered by path when using the project initializer.
+
+If for any reason you need custom registration, you may do so using `worker.registerActivities`:
+
+```ts
+// worker/run.ts
+import * as impl from './implementation/greeter'; // Non-standard path
+
+worker.registerActivities({ '@activities/greeter': impl });
```

Comment:
its not clear what are the rules around this naming schema `@activities/greeter`. 

1. can i name it something different? like `greeter`? what do we do when there are 2 activities with the same name?

2. using object syntax - do we plan to handle this?

```ts
import * as impl from './implementation/greeter'; // Non-standard path
import * as impl2 from './implementation/greeter2'; // Non-standard path
worker.registerActivities({ 
	'@activities/greeter': impl,
	'@activities/greeter2': impl2,
});
```

Reply:
that didn't fully answer my question - what are the rules of naming the key?  Is it any string goes?

Reply:
The way it works is that during import from a workflow we:
1. Check if an activities implementation is registered for this import specifier
2. If so inject implementation
3. Otherwise load normal ES module

So basically, any string goes, you directly impact the import system.